messaging/src/main/scala/uk/ac/wellcome/messaging/worker/Worker.scala

```diff
@@ -1,40 +1,40 @@
 package uk.ac.wellcome.messaging.worker
 
-import java.time.Instant
-
 import uk.ac.wellcome.messaging.worker.models.{Completed, Retry, WorkCompletion}
-import uk.ac.wellcome.messaging.worker.monitoring.MonitoringClient
 import uk.ac.wellcome.messaging.worker.steps.{
   MessageProcessor,
   MonitoringProcessor
 }
 
 import scala.concurrent.{ExecutionContext, Future}
 
-trait Worker[Message, Work, Summary, Action]
-    extends MessageProcessor[Message, Work, Summary]
-    with MonitoringProcessor {
+trait Worker[Message, Work, MonitoringContext, Summary, Action]
+    extends MessageProcessor[Message, Work, Summary] {
 
   type Processed = Future[(Message, Action)]
 
   implicit val ec: ExecutionContext
-  implicit val mc: MonitoringClient
 
   type Completion = WorkCompletion[Message, Summary]
   type MessageAction = Message => (Message, Action)
 
   protected val retryAction: MessageAction
   protected val completedAction: MessageAction
 
-  protected def processMessage(message: Message): Processed =
+  final def processMessage(message: Message)(
+    implicit monitoringProcessor: MonitoringProcessor[Message,
+                                                      MonitoringContext])
+    : Processed =
     work(message).map(completion)
 
-  private def work(message: Message): Future[Completion] = {
-    val startTime = Instant.now
-
+  private def work(message: Message)(
+    implicit monitoringProcessor: MonitoringProcessor[Message,
+                                                      MonitoringContext])
+    : Future[Completion] = {
     for {
+      context <- monitoringProcessor.recordStart(message)
       summary <- process(message)
-      monitor <- record(startTime, summary)
+      monitor <- monitoringProcessor.recordEnd(message, context, summary)
```

Comment:
It seems to me like if an exception is raised inside the processing of the message, we won't `recordEnd`. Is that correct, and is that the desired behaviour?

Reply:
No: `process` returns a `Future[Result[Summary]]` where `Result` encapsulates whether the ``Future` succeeded or failed. Even if there is an exception inside `process` it should be caught and encapsulated

Reply:
Oh yep, I'm guessing you mean [here](https://github.com/wellcomecollection/scala-messaging/blob/14b5fecf362220460d8ec2837ce29f0e37eed773/messaging/src/main/scala/uk/ac/wellcome/messaging/worker/steps/MessageProcessor.scala#L22)